zetaclient/chains/base/observer_inbound_backlog.go

```diff
@@ -0,0 +1,107 @@
+package base
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	zetaerrors "github.com/zeta-chain/node/pkg/errors"
+	crosschaintypes "github.com/zeta-chain/node/x/crosschain/types"
+	"github.com/zeta-chain/node/zetaclient/logs"
+)
+
+// GetInboundTrackersWithBacklog returns combined inbound trackers from zetacore and the backlog.
+func (ob *Observer) GetInboundTrackersWithBacklog(ctx context.Context) ([]crosschaintypes.InboundTracker, error) {
+	trackers, err := ob.ZetacoreClient().GetInboundTrackersForChain(ctx, ob.chain.ChainId)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get inbound trackers")
+	}
+
+	// create a map to look up trackers by inbound hash
+	trackersSet := make(map[string]bool)
+	for _, tracker := range trackers {
+		trackersSet[strings.ToLower(tracker.TxHash)] = true
+	}
+
+	ob.mu.Lock()
+	defer ob.mu.Unlock()
+
+	// combine trackers with the backlog and skip those that are already finalized
+	finalizedBallots := make([]string, 0)
+	for ballot, tracker := range ob.failedInboundBacklog {
```

Comment:
For both trackers and missed inbound, we should have a limit of how many are iterated in a single round